modules/_labs/select/react/lib/Select.tsx

```diff
@@ -0,0 +1,787 @@
+    ignoreDisabled: boolean = true
+  ): number => {
+    const childrenArray = React.Children.toArray(this.props.children) as React.ReactElement<
+      SelectOptionProps
+    >[];
+
+    for (let i = startIndex; i < endIndex; i++) {
+      const label = this.optionLabels[i].toLowerCase();
+      if (label.indexOf(startString.toLowerCase()) === 0) {
+        if (!ignoreDisabled || (ignoreDisabled && !childrenArray[i].props.disabled)) {
+          return i;
+        }
+      }
+    }
+
+    return -1;
+  };
+
+  private toggleMenu = (show: boolean): void => {
+    if (show) {
+      this.setState(
+        {
+          focusedOptionIndex: this.state.selectedOptionIndex,
+          isMenuHidden: false,
+        },
+        () => {
+          // Shift focus to the menu
+          if (this.menuRef.current) {
+            this.menuRef.current.focus();
+          }
+        }
+      );
+    } else {
+      this.setState({isMenuHiding: true});
+
+      this.removeMenuTimer = setTimeout(() => {
+        this.setState(
+          {
+            isMenuHiding: false,
+            isMenuHidden: true,
+          },
+          () => {
+            // Shift focus back to the button
+            if (this.buttonRef.current) {
+              this.buttonRef.current.focus();
+            }
+          }
+        );
+      }, toggleMenuAnimationDuration);
+    }
+  };
+
+  // Menu may not be interacted with while it is hiding
+  private isMenuInteractive = (): boolean => {
+    return !this.state.isMenuHiding;
+  };
+
+  // Code inspired by: https://stackoverflow.com/a/46012210
+  // In order for Select to be usable as a controlled component, we
+  // need to programatically change the value of the SelectInput
+  // in such a way that triggers its change event
+  private updateInput = (value: string): void => {
+    if (this.inputRef && this.inputRef.current) {
+      const nativeInputValue = Object.getOwnPropertyDescriptor(
+        Object.getPrototypeOf(this.inputRef.current),
+        'value'
+      );
+      if (nativeInputValue && nativeInputValue.set) {
+        nativeInputValue.set.call(this.inputRef.current, value);
+      }
+
+      let event: Event;
+      if (typeof Event === 'function') {
+        // Modern browsers
+        event = new Event('input', {bubbles: true});
+      } else {
+        // IE 11
+        event = document.createEvent('Event');
+        event.initEvent('input', true);
+      }
+
+      this.inputRef.current.dispatchEvent(event);
+    }
+  };
+
+  // TODO: Move code for scrollIntoViewIfNeeded to a centralized place.
+  // TODO: This code also has minor issues (sometimes it scrolls
+  // unnecessarily, the centerIfNeeded doesn't always center)
+  // Lifted from https://gist.github.com/hsablonniere/2581101
+  // This scrolling behavior is preferable even to the WebKit-proprietary
+  // scrollIntoViewIfNeeded method.
+  private scrollIntoViewIfNeeded = (elem: HTMLElement, centerIfNeeded = true): void => {
```

Comment:
What would be the recommended location for this scrolling code? I've pasted it directly into the Select component for now, but it should probably be placed elsewhere for possible reuse. I can also look into finding a package which provides the same functionality if that's preferable.

FYI: this function is used to keep the focused option in view when you use the keyboard to navigate a scrollable menu, and to scroll the selected option into view when you activate a scrollable menu.

Reply:
If the component is a functional component, this logic could be done as a React hook

Reply:
Implementing this as a hook in a later PR.